paddle/fluid/operators/math/concat.cu

```diff
@@ -0,0 +1,280 @@
+                                 const int input_col, const int output_cols,
+                                 CUDADeviceArray<T*> outputs) {
+  int tid_x = blockIdx.x * blockDim.x + threadIdx.x;
+  float inv_input_col = 1.0 / input_col;
+  for (; tid_x < input_col; tid_x += blockDim.x * gridDim.x) {
+    int split = tid_x * inv_input_col;
+    int in_offset = tid_x - split * input_col;
+    T* output_ptr = outputs.data[split];
+    int tid_y = blockIdx.y * blockDim.y + threadIdx.y;
+    for (; tid_y < input_row; tid_y += blockDim.y * gridDim.y)
+      output_ptr[tid_y * output_cols + in_offset] =
+          input[tid_y * input_col + tid_x];
+  }
+}
+
+/*
+ * All tensors' dimension should be the same.
+ */
+template <typename T>
+class ConcatFunctor<platform::CUDADeviceContext, T> {
+ public:
+  void operator()(const platform::CUDADeviceContext& context,
+                  const std::vector<framework::Tensor>& input, const int axis,
+                  framework::Tensor* output) {
+    // assume the the max size of input is less than 8 and see the performance
+    // save origin dim
+    int num = input.size();
+    PADDLE_ENFORCE_LT(num, MaxSize, "input number should be less than %d",
+                      MaxSize);
+    // get the matrix size
+    int rows = 1;
+    auto dim_0 = input[0].dims();
+    for (int i = 0; i < axis; ++i) {
+      rows *= dim_0[i];
+    }
+    int cols = input[0].numel() / rows;
+    int out_rows = rows, out_cols = 0;
+    bool sameShape = true;
+
+    CUDADeviceArray<const T*> inputs_data;
+    CUDADeviceArray<int> inputs_cols;
+    inputs_data.size = num;
+    inputs_cols.size = num + 1;
+    inputs_cols.data[0] = 0;
+    // reshape to matrix
+    // check input shape is valid
+    for (int i = 0; i < num; ++i) {
+      int t_cols = input[i].numel() / rows;
+      if (sameShape) {
+        if (t_cols != cols) sameShape = false;
+      }
+      out_cols += t_cols;
+      inputs_cols.data[i + 1] = out_cols;
+      inputs_data.data[i] = input[i].data<T>();
+    }
+
+    // computation
+    // set the thread block and grid according to CurrentDeviceId
+    const int kThreadsPerBlock = 1024;
+    int block_cols = std::min(out_cols, kThreadsPerBlock);
+    int block_rows = std::max(kThreadsPerBlock / block_cols, 1);
+    dim3 block_size = dim3(block_cols, block_rows, 1);
+
+    int dev_id = paddle::platform::GetCurrentDeviceId();
+    int multi_process = paddle::platform::GetCUDAMultiProcessors(dev_id);
+    int max_threads_per_mp =
+        paddle::platform::GetCUDAMaxThreadsPerMultiProcessor(dev_id);
+    int max_threads = multi_process * max_threads_per_mp;
+    int max_blocks = std::max(max_threads / kThreadsPerBlock, 1);
+
+    int grid_cols =
+        std::min((out_cols + block_cols - 1) / block_cols, max_blocks);
+    int grid_rows =
+        std::min(max_blocks / grid_cols, std::max(out_rows / block_rows, 1));
+    dim3 grid_size = dim3(grid_cols, grid_rows, 1);
```

Comment:
First of all, this calculation is correct. My only concern is that whether we truly need the information of **Multi-Processor**. 

In fact, I have tested the occupancy in the Titan X(pascal arch) with different grid size, block size,  finally find that if the block size >= 256, the grid size use upper bound of input size, I will get near 100% occupancy rate in my test. 

In a word, if the block size has been >= 256(8 times of warp size) the GPU occupancy and performance will be ok.

My test code is modified based https://github.com/zchee/cuda-sample/blob/master/0_Simple/simpleOccupancy/simpleOccupancy.cu.

Reply:
Furthermore, if we really need the attribute of GPU Device, we can store this information in DeviceContext. It is device related and will save a lot of heavy overhead device query.

Reply:
> the grid size use upper bound of input size

Please remind that grid's size is not unlimited. Assuming that the input is `x1` and `x2` and their shape is (100,1000000), axis is 0, the kernel will fail, you can have a try.

> get near 100% occupancy rate 

My experience is that, in some case, 100% occupancy doesn't mean the kernel is efficient. My test code is https://gist.github.com/chengduoZH/bc20fa8c12f8f045b74240dcdad41c84 . I also do some experiment and past the result in the comment.

> Furthermore, if we really need the attribute of GPU Device, we can store this information in DeviceContext. It is device related and will save a lot of heavy overhead device query.

I agreed that.

Reply:
> the grid size use upper bound of input size

Please remind that grid's size is not unlimited. Assuming that the input is `x1` and `x2` and their shape is (100,1000000), axis is 0, the kernel will fail, you can have a try.

> get near 100% occupancy rate 

My experience is that, in some case, 100% occupancy doesn't mean the kernel is efficient. My test code is https://gist.github.com/chengduoZH/bc20fa8c12f8f045b74240dcdad41c84 . I also do some experiment and past the result in the comment.

> Furthermore, if we really need the attribute of GPU Device, we can store this information in DeviceContext. It is device related and will save a lot of heavy overhead device query.

I agreed that.